Add a contenthost_factory to perform post-deploy actions

JacobCallahan opened the pull request:
The intent of this change is to enable us to add post-deploy actions when requesting specific types of content hosts, like fips-enabled. Users can define a set of host_post_configs in conf/content_host.yaml file. They can then add the post config action under the host definition.

## Summary by Sourcery

Introduce a reusable content host factory that centralizes host checkout and optional post-deploy configuration, and update fixtures and upgrade tests to use it while improving network defaults and host cleanup.

New Features:
- Add a contenthost_factory context manager to create and optionally post-configure content hosts based on settings-defined post-config profiles.
- Add a host_post_config helper to run configurable post-deploy broker actions on one or more content hosts.
- Introduce a registered_hosts fixture that provisions and registers multiple content hosts to Satellite based on the current setup.

Enhancements:
- Default content host network configuration to settings.content_host.network_type when not explicitly provided in fixture parameters.
- Refactor existing content host fixtures to rely on the shared contenthost_factory for consistent provisioning behavior, including _count-based multi-host scenarios and Satellite-based custom hosts.
- Adjust CentOS and Oracle host fixtures to derive deploy_network_type from the server IPv4/IPv6 configuration.
- Update the sat_upgrade_chost fixture to use the common contenthost_factory and UBI8 container host configuration instead of directly invoking Broker.

Tests:
- Add explicit cleanup of the post-upgrade RHEL client in the repository upgrade test via a finalizer that checks in the host through Broker.

JacobCallahan committed:
Add a contenthost_factory to perform post-deploy actions

The intent of this change is to enable us to add post-deploy actions
when requesting specific types of content hosts, like fips-enabled.
Users can define a set of host_post_configs in conf/content_host.yaml
file. They can then add the post config action under the host
definition.

jyejare commented:
@JacobCallahan I think the approach shown in https://github.com/SatelliteQE/robottelo/pull/16903/files is much more straightforward and has fewer changes needed to make this change today and in future as well.

Though I agree that the post deploy configs are not visible within the module and happen behind the scenes but its not impossible to find those. Hense I sold on the idea of https://github.com/SatelliteQE/robottelo/pull/16903#discussion_r1839969056 , else I too had a similar structure defined before changing my mind.

CC: @ogajduse

pcreech commented on pytest_fixtures/core/contenthosts.py:
```diff
@@ -41,93 +43,115 @@ def host_conf(request):
+        for key, val in broker_args.items():
+            if "{" in val:
+                broker_args[key] = val.format(host=host)
+        Broker(**broker_args).execute()
```
This is might be a random silly question, but it doesn't seem obvious on the surface to me upon an initial quick read.

My brain got caught up on the "why are we telling broker to take an action here, when it wasn't taking action before"

So in digging further, I'm inferring the following:

With the new approach, we now have actions we have to actually take time to execute.  In this case, either the 'enable-fips' workflow, or the 'enable-fapolicyd' workflow, as part of the 'factory<=>builder" relationship we're introducing.  When we encounter this, we then tell broker to take these specific actions before moving on to the next thing, which in this case would be yeilding the host back up to whatever called the contenthost_factory.

Does that seem generally correct?

JacobCallahan replied:
that's exactly right! it's like making sure the manufacturer puts all the extras on your vehicle before it leaves the factory.

JacobCallahan committed:
Refactor contenthost_factory and address review comments

- Fix host_post_config to copy broker_args to avoid side effects in loops.
- Add error handling and documentation to host_post_config.
- Update conf/content_host.yaml.template to use correct {host.name} placeholder.
- Fix sat_upgrade_chost fixture to correctly yield the host object from the factory context manager.

ogajduse reviewed (changes requested):
LGTM overall. Two changes need to be made before merging.

JacobCallahan committed:
Refactor content host fixture parameter handling

Refactoring:
- Standardize parameter passing to `contenthost_factory` across various fixtures, including `centos_host`, `oracle_host`, `external_puppet_server`, `sat_upgrade_chost`, and `custom_host`.
- Transition from directly mutating `request.param` to passing parameters as explicit keyword arguments to `contenthost_factory`.
- For `custom_host`, implement copying of `request.param` to prevent side effects, extract specific parameters (`deploy_rhel_version`, `workflow`), and pass any remaining custom parameters using `**kwargs`.
- Remove redundant `deploy_network_type` configuration from `centos_host` and `oracle_host` fixtures.

Fixes:
- Correct `sat_upgrade_chost` to use RHEL8 container host settings, resolving a previous reference to UBI8.

Documentation:
- Update the docstring for `sat_upgrade_chost` to accurately reflect its purpose and remove an outdated "leaks" comment.

vsedmik reviewed (commented):
Looks good to me, just proposed two removals if they don't feel too unrelated to the PR subject.

vsedmik commented on pytest_fixtures/core/contenthosts.py:
```diff
@@ -47,90 +49,136 @@ def host_conf(request):
     """A fixture for use with unittest classes. Provides a rhel7 Content Host object"""
-    with Broker(**host_conf(request), host_class=ContentHost) as host:
+    with contenthost_factory(request=request) as host:
         yield host
```
Hmm, okay. We could omit the extraction from the `6.16.z` auto cherry-pick PR (I can do it if you want me to).
Although I don't expect we would be executing the suite against `6.16.z` anymore, I don't like breaking the `6.16.z` branch by that removal even more.

ogajduse commented:
@vsedmik, how are you feeling about merging this one?

vsedmik commented:
> @vsedmik, how are you feeling about merging this one?

Quite comfortable, @ogajduse
